### PR TITLE
Initial userdev plugin implementation

### PR DESCRIPTION
**Pull request by jpenilla**
Implements a userdev plugin, which allows for building against Mojang-mapped Paper with sources attached in-IDE, with support for remapping compiled plugins to Spigot's runtime names.

Besides adding the userdev plugin, this PR also makes necessary changes to core and patcher to allow for Paper and forks to generate "paperweight development bundles", which are then published to a maven repository for the userdev plugin to consume as input.

See here for an example plugin using paperweight-userdev: https://github.com/jpenilla/paperweight-test-plugin

And here for the Paper changes: https://github.com/PaperMC/Paper/pull/6163

### CODE REVIEW

File: paperweight-lib/src/main/kotlin/tasks/RemapJar.kt [hunk 1]
```diff
@@ -22,14 +22,89 @@
...
 import kotlin.io.path.*
 import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.file.FileCollection
 import org.gradle.api.file.RegularFileProperty
 import org.gradle.api.provider.ListProperty
 import org.gradle.api.provider.Property
 import org.gradle.api.tasks.*
+import org.gradle.jvm.toolchain.JavaLauncher
+
+fun tinyRemapperArgsList(): List<String> {
```
Comment (Proximyst) [inline paperweight-lib/src/main/kotlin/tasks/RemapJar.kt:38]: Perhaps a (private?) val instead

File: paperweight-lib/src/main/kotlin/tasks/RunForgeFlower.kt [hunk 2]
```diff
@@ -34,6 +36,85 @@ import org.gradle.api.tasks.CompileClasspath
 import org.gradle.api.tasks.Internal
 import org.gradle.api.tasks.OutputFile
 import org.gradle.api.tasks.TaskAction
+import org.gradle.jvm.toolchain.JavaLauncher
+
+fun forgeFlowerArgList(): List<String> {
```
Comment (Proximyst) [inline paperweight-lib/src/main/kotlin/tasks/RunForgeFlower.kt:41]: Val here too

**Review by Proximyst — APPROVED** [review]
Cool! I assume you've tested this, so a check from me.

**Comment by OxKitsune** [comment]
What's blocking this from being merged?

**Comment by electronicboy** [comment]
main person working on this project is prioritising their own health over rushing around to get this stuff pulled in

**Comment by OxKitsune** [comment]
Ah that's fair. Thanks for the update :)

**Review by DenWav — APPROVED** [review]
Overall looks good.

I mostly only did sanity checks and familiarizing myself with the code here, but I trust you and your code, and it all looked sane. Just a couple tiny things that popped up.

I don't want my lack of time and energy right now preventing this from moving forward, so after addressing those things, feel free to merge. Would be cool ify ou could clean up the commit history a bit before merging, and please rebase and merge when you do.

File: paperweight-lib/src/main/kotlin/util/utils.kt [hunk 3]
```diff
@@ -36,14 +36,14 @@ import java.net.URI
...
+import java.security.NoSuchAlgorithmException
 import java.util.Collections
 import java.util.IdentityHashMap
 import java.util.Optional
 import java.util.concurrent.ThreadLocalRandom
-import kotlin.io.path.bufferedReader
-import kotlin.io.path.createDirectories
-import kotlin.io.path.exists
-import kotlin.io.path.name
+import kotlin.experimental.and
```
[the remark this answers is not in the document]
Comment (jpenilla) [reply paperweight-lib/src/main/kotlin/util/utils.kt:45]: It is, the import + the toHex method using it were moved from GeneratePaperclipPatch. By converting the byte to an int first we can avoid the experimental function though.

**Comment by JRoy** [comment]
@limbo-app ready

**Comment by JRoy** [comment]
I tried @jpenilla

**Review by JRoy — DISMISSED** [review]

**Review by Proximyst — APPROVED** [review]
limbo-app 🍞y

**Comment by MattBDev** [comment]
Woo!!

**Pull request merged**